### PR TITLE
[refactor] response utils

### PR DESCRIPTION
**Pull request by thibaultCha**
- cleaner, nicer response sender helper with unit tests
- utils is not a global variable anymore
- plugins and resolvers now have to **explicitly** stop a request from being proxied by setting `ngx.ctx.stop_phases` (old `ngx.ctx.error`. This value was also renamed because it's type (boolean) was confusing with this name and because it is not the responsability of the responses to set this value.
- Enforce `return` statement before `responses.send` functions, as advised by the nginx lua http module documentation.
- add tests for Server and Via headers on proxying success/failure
- move ngx_stub to tools for the CLI
- simplified is_port_open because it is not being run in the context of
  nginx anyways

### CODE REVIEW

File: kong/kong.lua [hunk 1]
```diff
@@ -189,11 +187,13 @@ function _M.exec_plugins_access()
 end
 
 -- Calls header_filter() on every loaded plugin
--- @return nil
 function _M.exec_plugins_header_filter()
   ngx.ctx.proxy_ended_at = timestamp.get_utc() -- Setting a property that will be available for every plugin
 
-  if not ngx.ctx.error then
+  if not ngx.ctx.stop_phases then
+    ngx.log(ngx.ERR, "IN HEADER_FILTER")
```
Comment (thefosk) [inline kong/kong.lua]: @thibaultCha maybe this line should be removed?

Comment (thibaultCha) [inline kong/kong.lua]: yepp

File: kong/kong.lua [hunk 2]
```diff
@@ -46,8 +47,7 @@ local function load_plugin_conf(api_id, consumer_id, plugin_name)
         name = plugin_name
       }
       if err then
-        ngx.log(ngx.ERR, tostring(err))
-        utils.show_error(500)
+        responses.send_HTTP_INTERNAL_SERVER_ERROR(err)
```
Comment (thibaultCha) [inline kong/kong.lua]: missing a return here to enforce the policy

File: kong/api/routes/base_controller.lua [hunk 3]
```diff
@@ -33,21 +32,18 @@ local function render_list_response(req, data, size)
   end
 end
 
-local function parse_dao_error(err)
-  local status
+local function send_dao_error_response(err)
   if err.database then
-    status = 500
-    ngx.log(ngx.ERR, tostring(err))
+    responses.send_HTTP_INTERNAL_SERVER_ERROR(err.message)
```
Comment (thibaultCha) [inline kong/api/routes/base_controller.lua]: and here

**Pull request merged**